File: implementations/php/pcf-dcp/src/DcpReader.php
<?php

declare(strict_types=1);

namespace Kduma\PCFDCP;

use Kduma\PCF\Consts as PcfConsts;
use Kduma\PCF\Container;
use Kduma\PCF\PartitionEntry;
use Kduma\PCF\Storage\StorageInterface;
use Kduma\PCF\TableBlockHeader;

/**
 * A reader for DCP containers layered over a PCF file. It works entirely through
 * the high-level {@see Container} API, so a DCP file written in trailer mode
 * reads back transparently.
 */
final class DcpReader
{
    private function __construct(
        private readonly Container $container,
    ) {
    }

    /** Open a PCF file for DCP-aware reading. */
    public static function open(StorageInterface $storage): self
    {
        return new self(Container::open($storage));
    }

    /** Borrow the underlying PCF container. */
    public function container(): Container
    {
        return $this->container;
    }

    /**
     * All top-level entries, in chain order.
     *
     * @return PartitionEntry[]
     */
    public function entries(): array
    {
        return $this->container->entries();
    }

    /**
     * The top-level DCP container entries.
     *
     * @return PartitionEntry[]
     */
    public function containers(): array
    {
        $out = [];
        foreach ($this->container->entries() as $e) {
            if ($e->partitionType === Consts::DCP_CONTAINER_TYPE) {
                $out[] = $e;
            }
        }

        return $out;
    }

    /** Parse the arena of a DCP container entry. */
    public function openArena(PartitionEntry $entry): Arena
    {
        if ($entry->partitionType !== Consts::DCP_CONTAINER_TYPE) {
            throw PcfDcpException::notADcpContainer();
        }

        return Arena::parse($this->container->readPartitionData($entry));
    }

    /**
     * Every inner partition across every DCP container, in file order.
     *
     * @return InnerLocation[]
     */
    public function innerPartitions(): array
    {
        $out = [];
        foreach ($this->containers() as $cont) {
            $arena = $this->openArena($cont);
            foreach ($arena->inners() as $info) {
                $out[] = new InnerLocation($cont->uid, $info);
            }
        }

        return $out;
    }

    /** Resolve a uid against the flattened set top-level ∪ inner (spec 2.1). */
    public function resolveUid(string $uid): Resolved
    {
        foreach ($this->container->entries() as $e) {
            if ($e->uid === $uid) {
                return Resolved::topLevel($e);
            }
        }
        foreach ($this->innerPartitions() as $loc) {
            if ($loc->info->uid === $uid) {
                return Resolved::innerPartition($loc);
            }
        }
        throw PcfDcpException::notFound();
    }

    /** Reconstruct an inner partition's logical content by uid. */
    public function readInner(string $uid): string
    {
        foreach ($this->containers() as $cont) {
            $arena = $this->openArena($cont);
            foreach ($arena->uids() as $u) {
                if ($u === $uid) {
                    return $arena->content($uid);
                }
            }
        }
        throw PcfDcpException::notFound();
    }

    /**
     * Full DCP-aware verification: PCF integrity, each inner Table Block's
     * table_hash, reconstruction length and (when algorithmic) data_hash, no
     * nested container, and file-wide uid uniqueness.
     */
    public function verify(): void
    {
        $this->container->verify();

        $seen = [];
        foreach ($this->container->entries() as $e) {
            $k = bin2hex($e->uid);
            if (isset($seen[$k])) {
                throw PcfDcpException::duplicateUid();
            }
            $seen[$k] = true;
        }

        foreach ($this->containers() as $cont) {
            $data = $this->container->readPartitionData($cont);
            self::verifyInnerTableHashes($data);

            $arena = Arena::parse($data);
            foreach ($arena->inners() as $info) {
                if ($info->partitionType === Consts::DCP_CONTAINER_TYPE) {
                    throw PcfDcpException::nestedContainer();
                }
                $k = bin2hex($info->uid);
                if (isset($seen[$k])) {
                    throw PcfDcpException::duplicateUid();
                }
                $seen[$k] = true;

                $content = $arena->content($info->uid);
                if (\strlen($content) !== $info->usedBytes) {
                    throw PcfDcpException::lengthMismatch($info->usedBytes, \strlen($content));
                }
                if (!$info->dataHashAlgo->verify($content, $info->dataHash)) {
                    throw PcfDcpException::hashMismatch();
                }
            }
        }
    }

    private static function verifyInnerTableHashes(string $arena): void
    {
        $header = DcpHeader::read($arena);
        $len = \strlen($arena);
        $off = $header->innerTableOffset;
        $budget = intdiv($len, PcfConsts::TABLE_HEADER_SIZE) + 1;
        while ($off !== 0) {
            if ($budget === 0) {
                throw PcfDcpException::offsetOutOfRange();
            }
            --$budget;
            if ($off + PcfConsts::TABLE_HEADER_SIZE > $len) {
                throw PcfDcpException::offsetOutOfRange();
            }
            $h = TableBlockHeader::fromBytes(substr($arena, $off, PcfConsts::TABLE_HEADER_SIZE));
            $entries = [];
            for ($i = 0; $i < $h->partitionCount; ++$i) {
                $eo = $off + PcfConsts::TABLE_HEADER_SIZE + $i * PcfConsts::ENTRY_SIZE;
                if ($eo + PcfConsts::ENTRY_SIZE > $len) {
                    throw PcfDcpException::offsetOutOfRange();
                }
                $entries[] = PartitionEntry::fromBytes(substr($arena, $eo, PcfConsts::ENTRY_SIZE));
            }
            if ($h->tableHashAlgo->verifies()) {
                $computed = TableBlockHeader::computeTableHash($h->tableHashAlgo, $h->nextTableOffset, $entries);
                $n = $h->tableHashAlgo->digestLen();
                if (substr($computed, 0, $n) !== substr($h->tableHash, 0, $n)) {
                    throw PcfDcpException::hashMismatch();
                }
            }
            $off = $h->nextTableOffset;
        }
    }
}

File: implementations/dotnet/pcf-dcp/src/Pcf.Dcp/DcpReader.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using Pcf;

namespace Pcf.Dcp;

/// <summary>An inner partition together with the container that holds it.</summary>
public sealed class InnerLocation
{
    /// <summary>uid of the enclosing DCP container partition.</summary>
    public byte[] ContainerUid { get; set; }

    /// <summary>The inner partition's metadata and extents.</summary>
    public InnerInfo Info { get; set; }
}

/// <summary>The result of resolving a uid against top-level ∪ inner (spec 2.1).</summary>
public sealed class Resolved
{
    /// <summary>Whether the uid resolved to a top-level PCF partition.</summary>
    public bool IsTopLevel { get; set; }

    /// <summary>The top-level entry (when <see cref="IsTopLevel"/> is true).</summary>
    public PartitionEntry Entry { get; set; }

    /// <summary>The inner partition location (when <see cref="IsTopLevel"/> is false).</summary>
    public InnerLocation Inner { get; set; }
}

/// <summary>
/// A reader for DCP containers layered over a PCF file. It works entirely
/// through the high-level <see cref="Container"/> API, so a DCP file written in
/// trailer mode reads back transparently.
/// </summary>
public sealed class DcpReader
{
    private readonly Container _c;

    private DcpReader(Container c)
    {
        _c = c;
    }

    /// <summary>Open a PCF file for DCP-aware reading.</summary>
    public static DcpReader Open(Stream storage) => new DcpReader(Container.Open(storage));

    /// <summary>Borrow the underlying PCF container.</summary>
    public Container Container => _c;

    /// <summary>All top-level entries, in chain order.</summary>
    public List<PartitionEntry> Entries() => _c.Entries();

    /// <summary>The top-level DCP container entries.</summary>
    public List<PartitionEntry> Containers()
    {
        var outList = new List<PartitionEntry>();
        foreach (var e in _c.Entries())
        {
            if (e.PartitionType == Constants.DcpContainerType)
            {
                outList.Add(e);
            }
        }
        return outList;
    }

    /// <summary>Parse the arena of a DCP container entry.</summary>
    public Arena OpenArena(PartitionEntry entry)
    {
        if (entry.PartitionType != Constants.DcpContainerType)
        {
            throw PcfDcpException.NotADcpContainer();
        }
        return Arena.Parse(_c.ReadPartitionData(entry));
    }

    /// <summary>Every inner partition across every DCP container, in file order.</summary>
    public List<InnerLocation> InnerPartitions()
    {
        var outList = new List<InnerLocation>();
        foreach (var cont in Containers())
        {
            var arena = OpenArena(cont);
            foreach (var info in arena.Inners())
            {
                outList.Add(new InnerLocation { ContainerUid = (byte[])cont.Uid.Clone(), Info = info });
            }
        }
        return outList;
    }

    /// <summary>Resolve a uid against the flattened set top-level ∪ inner (spec 2.1).</summary>
    public Resolved ResolveUid(byte[] uid)
    {
        foreach (var e in _c.Entries())
        {
            if (Arena.BytesEqual(e.Uid, uid))
            {
                return new Resolved { IsTopLevel = true, Entry = e };
            }
        }
        foreach (var loc in InnerPartitions())
        {
            if (Arena.BytesEqual(loc.Info.Uid, uid))
            {
                return new Resolved { IsTopLevel = false, Inner = loc };
            }
        }
        throw PcfDcpException.NotFound();
    }

    /// <summary>Reconstruct an inner partition's logical content by uid.</summary>
    public byte[] ReadInner(byte[] uid)
    {
        foreach (var cont in Containers())
        {
            var arena = OpenArena(cont);
            foreach (var u in arena.Uids())
            {
                if (Arena.BytesEqual(u, uid))
                {
                    return arena.Content(uid);
                }
            }
        }
        throw PcfDcpException.NotFound();
    }

    /// <summary>
    /// Full DCP-aware verification: PCF integrity, each inner Table Block's
    /// table_hash, reconstruction length and (when algorithmic) data_hash, no
    /// nested container, and file-wide uid uniqueness.
    /// </summary>
    public void Verify()
    {
        _c.Verify();

        var seen = new HashSet<string>();
        foreach (var e in _c.Entries())
        {
            if (!seen.Add(Hex(e.Uid)))
            {
                throw PcfDcpException.DuplicateUid();
            }
        }

        foreach (var cont in Containers())
        {
            var data = _c.ReadPartitionData(cont);
            VerifyInnerTableHashes(data);

            var arena = Arena.Parse(data);
            foreach (var info in arena.Inners())
            {
                if (info.PartitionType == Constants.DcpContainerType)
                {
                    throw PcfDcpException.NestedContainer();
                }
                if (!seen.Add(Hex(info.Uid)))
                {
                    throw PcfDcpException.DuplicateUid();
                }
                var content = arena.Content(info.Uid);
                if ((ulong)content.Length != info.UsedBytes)
                {
                    throw PcfDcpException.LengthMismatch((long)info.UsedBytes, content.Length);
                }
                if (!info.DataHashAlgo.Verify(content, info.DataHash))
                {
                    throw PcfDcpException.HashMismatch();
                }
            }
        }
    }

    private static void VerifyInnerTableHashes(byte[] arena)
    {
        DcpHeader header = DcpHeader.Read(arena);
        ulong off = header.InnerTableOffset;
        int budget = arena.Length / (int)Pcf.Constants.TableHeaderSize + 1;
        while (off != 0)
        {
            if (budget == 0)
            {
                throw PcfDcpException.OffsetOutOfRange();
            }
            budget -= 1;
            int baseOff = checked((int)off);
            if (baseOff + (int)Pcf.Constants.TableHeaderSize > arena.Length)
            {
                throw PcfDcpException.OffsetOutOfRange();
            }
            var hb = new byte[(int)Pcf.Constants.TableHeaderSize];
            Buffer.BlockCopy(arena, baseOff, hb, 0, hb.Length);
            var h = TableBlockHeader.FromBytes(hb);
            var entries = new List<PartitionEntry>(h.PartitionCount);
            for (int i = 0; i < h.PartitionCount; i++)
            {
                int eo = baseOff + (int)Pcf.Constants.TableHeaderSize + i * (int)Pcf.Constants.EntrySize;
                if (eo + (int)Pcf.Constants.EntrySize > arena.Length)
                {
                    throw PcfDcpException.OffsetOutOfRange();
                }
                var eb = new byte[(int)Pcf.Constants.EntrySize];
                Buffer.BlockCopy(arena, eo, eb, 0, eb.Length);
                entries.Add(PartitionEntry.FromBytes(eb));
            }
            if (h.TableHashAlgo.Verifies())
            {
                var computed = TableBlockHeader.ComputeTableHash(h.TableHashAlgo, h.NextTableOffset, entries);
                int n = h.TableHashAlgo.DigestLen();
                for (int i = 0; i < n; i++)
                {
                    if (computed[i] != h.TableHash[i])
                    {
                        throw PcfDcpException.HashMismatch();
                    }
                }
            }
            off = h.NextTableOffset;
        }
    }

    private static string Hex(byte[] b)
    {
        var sb = new StringBuilder(b.Length * 2);
        foreach (var x in b)
        {
            sb.Append(x.ToString("x2"));
        }
        return sb.ToString();
    }
}
